### DISCUSSION TITLE
Calling a route handler from a route handler because of caching

### DISCUSSION THREAD

**Discussion by funes79** [vercel/next.js#58682] [App Router] [answered]
I expect this is an anti-pattern, but I could not find any way how to cache for 30seconds the API (AppSync) call on server side. In localhost it works nicely, but in docker deployment it returns Not found for http://localhost:3000/api/graphql,  or http://127.0.0.1/api/graphl.

I tried to put `export const revalidate = 30;` to  the `api/todos/route.js` but it had not effect. 

Steps to reproduce:

1. Create two route handlers /api/todos and api/graphql. The first is for the client component to call. The second will be called from the server by the first one, and revalidate (caching will be applied). The `api/graphql` takes the Todo resources and fetches from AppSync. To avoid reading the Todos every time, the client calls `/api/todos`  what is then doing a `fetch` method to get the `api/graphql`
2. Create a home page with client component doing a fetch on the api

```
// app/api/todos/route.js
import { serverClient } from '@/utils/server-utils';
import { listTodos } from '@/src/graphql/queries';

export async function GET(request) {
  let data = {
    message: 'Hello from the API v2',
  };
  console.log('Going to call another Route from Route');
  let url = `http://localhost:3000/api/graphql`;
  const nested = await fetch(url, {
// we need to pass the request.headers, so that Cognito tokens are received and the nested API can fetch the data
    headers: request.headers,
    next: { revalidate: 30 },
  });
  const nestedData = await nested.json();
  data.todos = nestedData?.data?.todos;
  return Response.json({ data });
} 
```

```
// app/api/graphql/route.js
import { serverClient } from '@/utils/server-utils';
import { listTodos } from '@/src/graphql/queries';

export async function GET(request) {
  // lets log out if we have cookies
  console.log('request api/graphql', request?.cookies?.getAll());
  let data;
  try {
    const todos = await serverClient.graphql({
      query: listTodos,
    });
    data = {
      message: 'Hello from nested the API',
      todos: todos.data.listTodos.items,
    };
  } catch (err) {
    return Response.json({}, { status: 404 });
  }
  return Response.json({ data });
}
```

Client component:

```
function APITester() {
  const [msg, setMsg] = React.useState('');
  const test = async () => {
    setMsg('Calling API...');
    let url = '/api/todos';
    // this revalidate does not seems to have any effect, so I decided to cache on server side
    const response = await fetch(url, {
      next: { revalidate: 15 },
    });
    const resp = await response.json();
    setMsg( `${resp.data.message} cnt: ${JSON.stringify(resp?.data?.todos?.length)}` );
  };
  return (
    <>
      <div>APITester</div>
      <div className="flex">
        <button className="btn" onClick={() => test()}>
          Call router handler (nested)
        </button>
      </div>
      <p>API response: {msg}</p>
    </>
  );
}
```

**Comment by matthewraaff** [comment]
You can use the Cache-Control header to achieve this. When returning a new response, add it to the Response object.

```js
  return new Response(JSON.stringify({ data }), {
    headers: {
      'Cache-Control': 's-maxage=30',
      'Content-Type': 'application/json',
    },
  });
```

**Comment by funes79** [comment]
Thanks @matthewraaff . But this setting solves the issue partially. The client calls the api, the API returns the data (I see in the headers the cache-control) - lets say 4 items. Then I create a new item via client component I call the API and it returns 4 items (because it is cached) but immediately after hitting the API returns 5. So there is no delay, I would expect that the first call creates a cache, and for 30 seconds I got stale data. But it seems that every time I add new record, and call the API, I got stale, and the second call returns the correct count.

So I assume the first cached call triggers cache refresh, but returns stale data. And the 2nd call shortly after that, it returns fresh data. 

My expectation was that if I set 30s or 3600s then for that duration the data is never refreshed.

**Comment by matthewraaff** [comment, marked as answer]
@funes79 Ah, sorry for misunderstanding.

You could try running a basic caching implementation on the server instead of completely relying on HTTP headers - This could probably be done by assigning the current time to a variable when a GET request has been made and checking if it's 30+ seconds after the last request, then fetching new data rather than using the older data. You could also start an asynchronous task that runs every 30 seconds to fetch the newer data, though this might be a bit excessive and store a lot of redundant data temporarily.

I've written a small (untested) implementation [here](https://gist.github.com/matthewraaff/50d1440407cfa9dc23bd6d6dcdc1138f). I've also added the stale-while-revalidate header, as it might change how it works slightly, but if it still doesn't work, try removing it.

I hope this helps :)

**Comment by funes79** [comment]
@matthewraaff thanks for the reply. Its a nice workaround, and I tested it works like charm. 👍
